### PR TITLE
Feature/LJI-17-packages-filters

### PR DESCRIPTION
**Pull request by anddevdev**
This PR introduces the following features for the Pacakges:

1.Filtering packages:

-Allows filtering of packages by ame, status, start date, end date and payroll frequency
-Multiple filters can be applied simultaneously

2.Sorting Pacakges:

- Enables sorting of packages by name, status, payroll frequency, start date and end date (Ascending and descending orders)
- Default sorting is enabled by name in ascending order.

3.Pagination:

- Default pagination is set to 10 entities per page
New endpoint:
Added a new endpoint /v1/pacakges/filtered for handling filtered, sorted, and paginated requests.
Added new requests to Packages collection in Postman to test the added features.

### CODE REVIEW

**Review by SavaLX — APPROVED** [review]

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java [hunk 1]
```diff
@@ -39,6 +43,33 @@ public class PackageIntegrationTest {
   @MockBean
```
[the remark this answers is not in the document]
Comment (anddevdev) [reply src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java:48]: Done

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java [hunk 2]
```diff
@@ -78,48 +139,40 @@ void testGetPackageById() throws Exception {
...
-    String packageJson = objectMapper.writeValueAsString(packageDTO);
+    UUID packageId = UUID.randomUUID();
+    PackageDTO originalPackageDTO = easyRandom.nextObject(PackageDTO.class);
+    originalPackageDTO.setId(packageId);
+    PackageEntity packageEntity = PackageMapper.INSTANCE.dtoToEntity(originalPackageDTO);
 
-    Mockito.when(packageService.updatePackage(Mockito.any())).thenReturn(packageDTO);
+    Mockito.when(packageRepository.findById(eq(packageId))).thenReturn(Optional.of(packageEntity));
+    Mockito.when(packageRepository.save(any(PackageEntity.class))).thenReturn(packageEntity);
+    Mockito.when(packageService.updatePackage(any(PackageDTO.class))).thenReturn(originalPackageDTO);
```
Comment (NormantasN) [inline src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java:150]: Refer to comment [#26 (comment)](https://github.com/Java-POT-Summer-24/pot-java-summer-24/pull/26#discussion_r1666566526) 
Refactor updatePackage tests accordingly

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java [hunk 3]
```diff
@@ -39,36 +48,88 @@ public class PackageIntegrationTest {
...
   private PackageService packageService;
 
+  @MockBean
+  private PackageRepository packageRepository;
+
+  private final EasyRandom easyRandom = new EasyRandom();
+
+  @Test
+  void testGetFilteredSortedPackages() throws Exception {
+    List<PackageEntity> packageEntities = easyRandom.objects(PackageEntity.class, 3)
```
Comment (NormantasN) [inline src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java:58]: You are using only packageDTOs, you can optimize by deleting packageEntities and creating DTOs as easyRandom.objects

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java [hunk 4]
```diff
@@ -39,36 +48,88 @@ public class PackageIntegrationTest {
...
-        .andExpect(jsonPath("$.name").value("Basic Health Package"));
+        .andExpect(jsonPath("$.name").value(packageDTO.getName()));
   }
 
   @Test
   void testGetAllPackages() throws Exception {
+    List<PackageEntity> packageEntities = easyRandom.objects(PackageEntity.class, 3).collect(Collectors.toList());
+    List<PackageDTO> packageDTOs = packageEntities.stream()
+        .map(PackageMapper.INSTANCE::entityToDto)
+        .collect(Collectors.toList());
```
Comment (NormantasN) [inline src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java:113]: You can optimize with .toList();

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageControllerTest.java [hunk 5]
```diff
@@ -46,8 +52,7 @@ void testGetAllPackages() {
     List<PackageDTO> packageList = easyRandom.objects(PackageDTO.class, 3).toList();
     when(packageService.getAllPackages()).thenReturn(packageList);
 
-    ResponseEntity<List<PackageDTO>> response = packageController.getAllPackages();
-    List<PackageDTO> result = response.getBody();
+    List<PackageDTO> result = packageController.getAllPackages();
 
     assertEquals(3, result.size());
```
Comment (NormantasN) [inline src/test/java/com/coherentsolutions/pot/insurance/packages/PackageControllerTest.java:57]: I believe you should consider a case when there are objects in the repository already, my suggestion:
assertEquals(packageRepository.count() + 3, result.size());

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageServiceTest.java [hunk 6]
```diff
@@ -16,10 +19,14 @@
...
-import org.springframework.http.ResponseEntity;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
 
 @ExtendWith(MockitoExtension.class)
-class PackageControllerTest {
+class PackageServiceTest {
```
Comment (SavaLX) [inline src/test/java/com/coherentsolutions/pot/insurance/packages/PackageServiceTest.java:29]: @InjectMocks - PackageService, @Mock - PackageRepository, call service method to get and verify the result and mock repository method

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java (hunk 2 above)
Comment (anddevdev) [reply src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java:133]: Done

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java (hunk 4 above)
Comment (anddevdev) [reply src/test/java/com/coherentsolutions/pot/insurance/packages/PackageIntegrationTest.java]: Done

File: src/test/java/com/coherentsolutions/pot/insurance/packages/PackageServiceTest.java (hunk 6 above)
Comment (anddevdev) [reply src/test/java/com/coherentsolutions/pot/insurance/packages/PackageServiceTest.java:32]: Done

**Pull request merged**